src/stream/src/executor/managed_state/join/mod.rs

```diff
@@ -151,149 +156,176 @@ impl<K: HashKey, S: StateStore> JoinHashMap<K, S> {
+
+/// Iterates
+pub(crate) struct IterJoinEntriesMut<'a, K: HashKey, S: StateStore> {
+    side_match: &'a mut JoinHashMap<K, S>,
+    side_update: &'a mut JoinHashMap<K, S>,
+    last_item: Option<IterType<K, S>>,
+    queue: tokio::sync::mpsc::UnboundedReceiver<IterType<K, S>>, /* TODO: change this to a
+                                                                  * reasonably
+                                                                  * sized bounded queue... */
+}
+
+impl<'a, K: HashKey, S: StateStore> IterJoinEntriesMut<'a, K, S> {
+    pub(crate) fn new(
+        side_match: &'a mut JoinHashMap<K, S>,
+        side_update: &'a mut JoinHashMap<K, S>,
+        keys: impl Iterator<Item = &'a K>,
+    ) -> Self {
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel::<IterType<K, S>>();
+        let new_iter = Self {
+            side_match,
+            side_update,
+            last_item: None,
+            queue: rx,
+        };
+
+        for key in keys {
+            let update_entry = new_iter.side_update.pop_cached(key);
+            let match_entry = if !key.has_null() {
+                new_iter.side_match.pop_cached(key)
+            } else {
+                None
+            };
+            let key = key.clone();
+            let match_table_info = new_iter.side_match.table_info.clone();
+            let update_table_info = new_iter.side_update.table_info.clone();
+            let tx = tx.clone();
+
+            tokio::spawn(async move {
+                let state = Self::get_key(
+                    key,
+                    match_entry,
+                    update_entry,
+                    match_table_info,
+                    update_table_info,
+                )
+                .await;
+                tx.send(state)
+                    .ok()
+                    .expect("Failed to send join entries for join key");
+            });
+        }
+
+        new_iter
+    }
+
+    pub(crate) async fn next(&mut self) -> Option<&mut IterType<K, S>> {
+        if let Some(last) = self.last_item.take() {
+            if let Some(matches) = last.1 {
+                self.side_match.push(last.0.clone(), matches);
+            }
+            self.side_update.push(last.0, last.2);
+        }
+        self.last_item = self.queue.recv().await;
+        self.last_item.as_mut()
+    }
+
+    pub(crate) async fn get_key(
+        key: K,
+        mut match_entry: Option<JoinEntryState<S>>,
+        update_entry: Option<JoinEntryState<S>>,
+        match_table_info: TableInfo<S>,
+        update_table_info: TableInfo<S>,
+    ) -> IterType<K, S> {
+        if !key.has_null() && match_entry.is_none() {
+            match_entry = JoinHashMap::<K, S>::fetch_cached_state(&key, &match_table_info)
+                .await
+                .ok()
+                .unwrap();
+        }
+        let update_entry = update_entry.unwrap_or_else(|| {
+            JoinHashMap::<K, S>::init_without_cache(&key, &update_table_info)
```

Comment:
This is different from the current behaviour. Currently, the entry will be `None` if storage does not contain the join key. So the current behaviour is to noop on deletion if there are no entries. For us, we are simply allowing the update to pass through to save on querying storage.